<a href="https://colab.research.google.com/github/Mohit03singh/ALMAX_PYTHON_COMPETITION/blob/main/Almax_python_pratice.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

question 13

In [ ]:
def plusOne(digits):
    n = len(digits)
    for i in range(n):
        idx = n - 1 - i
        if digits[idx]<9:
            digits[idx]+=1
            return digits
        else:
            digits[idx]=0
    return[1]+digits

digits = eval(input())

# Display the result after adding one
print(plusOne(digits))

[1,2,3]
[1, 2, 4]


In [ ]:
def titleToNumber(s):
    # Initialize the result
    result = 0
    char_map={chr(i):i-64 for i in range(65,91)}
    n = len(s)
    for i in range(n):
        pos=n-1-i
        cur_pos=s[pos]
        result +=char_map[cur_pos]*(26**i)
    return result
st = input()
print(titleToNumber(st))

ZY
701


In [ ]:
def isHappy(n):
    # Function to calculate the sum of squares of digits
    def get_next(number):
        total_sum=0
        while number>0:
            digit=number%10
            total_sum += digit * digit
            number//=10
        return total_sum
    seen = set()
    while n != 1 and n not in seen:

        seen.add(n)
        n=get_next(n)
    return n == 1
n = int(input())
print(isHappy(n))

19
True


In [ ]:
def isPowerOfTwo(n):
    if n == 0:
        return False
    return n & (-n) == n

# Prompt for user input
n = int(input())
if isPowerOfTwo(n):
    print('True')
else:
    print('False')

1
True


In [ ]:
def buildArray(nums):
    n=len(nums)
    ans=[0]*n

    for i in range(n):
        ans[i]=nums[nums[i]]

    return ans
lst = eval(input())
print(buildArray(lst))

[0,2,1,5,3,4]
[0, 1, 2, 4, 5, 3]


In [ ]:
def getConcatenation(nums):
    concatenated=nums+nums
    return concatenated

lst = eval(input())
print(getConcatenation(lst))

[1,2,1]
[1, 2, 1, 1, 2, 1]


In [ ]:
def numIdenticalPairs(nums):
    # Initialize the count of good pairs
    ans = 0

    n=len(nums)
    for i in range(n):
        for j in range(i+1,n):
            if nums[i]==nums[j] and i<j:
              ans+=1
    return ans
lst = eval(input())
print(numIdenticalPairs(lst))

[1,2,3,1,1,3]
4


In [ ]:
def shuffle(nums, n):
    result = [0] * (2 * n)
    for i in range(n):
        result[2*i]=nums[i]
        result[2*i+1]=nums[n+i]
    return result

lst = eval(input())
nm = int(input())

print(shuffle(lst, nm))


[2,5,1,3,4,7]
3
[2, 3, 5, 4, 1, 7]


In [ ]:
def numberOfEmployeesWhoMetTarget(hours, target):
    # Initialize the count of employees meeting the target
    count = 0
    for hour in hours:
        if hour>=target:
            count+=1
    return count
hours = eval(input())

# Prompt for the target hours
t = int(input())
print(numberOfEmployeesWhoMetTarget(hours, t))


[0,1,2,3,4]
2
3


In [ ]:
def maximumWealth(accounts):
    max_wealth_so_far = 0
    for account in accounts:
        curr_customer_wealth=sum(account)
        if curr_customer_wealth>max_wealth_so_far:
            max_wealth_so_far=curr_customer_wealth
    return max_wealth_so_far
ac = eval(input())

# Display the maximum wealth among all customers
print(maximumWealth(ac))

[[1,2,3],[3,2,1]]
6


In [ ]:
def kidsWithCandies(candies, extraCandies):
    maxCandies = max(candies)
    result = []
    for candy in candies:
        result.append(candy+extraCandies>=maxCandies)
    return result
cand = eval(input())
ec = int(input())

# Display the result indicating which kids can have the greatest number o candies
print(kidsWithCandies(cand, ec))

[2,3,5,1,3]
3
[True, True, True, False, True]


In [ ]:
def runningSum(nums):
    for i in range(1,len(nums)):
        nums[i] += nums[i - 1]
    return nums
lst = eval(input())
print(runningSum(lst))


[1,2,3,4]
[1, 3, 6, 10]


In [ ]:
def smallerNumbersThanCurrent(nums):
    sorted_nums=sorted(nums)
    map1={}
    for idx,num in enumerate(sorted_nums):
        if num not in map1:
            map1[num]=idx
    result = []
    result = [map1[num] for num in nums]
    return result
lst = eval(input())
print(smallerNumbersThanCurrent(lst))

[8,1,2,2,3]
[4, 0, 1, 1, 3]


In [ ]:
def Fibonacci(n):
    if n <=0:
        return "Incorrect input"
    if n==1:
        return 0
    if n==2:
        return 1
    if n>2:
        return Fibonacci(n-1)+Fibonacci(n-2)
n=eval(input())
print(Fibonacci(n))

10
34


In [ ]:
def finalValueAfterOperations(operations):
    result = 0
    for operation in operations:
        if operation =="++X" or operation =="X++":
            result+=1
        else:
            result-=1

    return result
l = eval(input())
print(finalValueAfterOperations(l))

["--X","X++","X++"]
1


In [ ]:
# Function to calculate the factorial of a number
def calculate_factorial():
    num = int(input())
    factorial = 1
    if num <0:
        print("Factorial can't be calculated for negative number")
    elif num ==0:
        print("Factorial of 0 is 1")
    else:
        for i in range(1,num+1):
            factorial*=i
        print("Factorial of", num, "is", factorial)

# Main function
calculate_factorial()

5
Factorial of 5 is 120


In [ ]:
# Function to check if a number is an Armstrong number
def check_armstrong_number():
    number = int(input())
    temp=number
    i=0
    while number >0:
        number//=10
        i+=1

    number=temp
    result=0
    while number > 0:
        digit = number % 10
        result += digit ** i
        number //= 10

    if temp ==result:
        print("Number is Armstrong")
    else:
        print("Number is not Armstrong")

# Call the check_armstrong_number function
check_armstrong_number()

153
Number is Armstrong


In [ ]:
def isPalindrome(x):
    if x <0:
        return False
    reversed_num = 0
    temp = x
    while temp>0:
        digit =temp%10
        reversed_num=reversed_num *10+digit
        temp//=10
    return x==reversed_num

# Prompt for user input
num = int(input())
print(isPalindrome(num))


121
True


In [ ]:
def prime(x, y):
    prime_list = []
    for i in range(x, y):
        if i == 0 or i == 1:
            continue
        is_prime=True
        for j in range(2,int(0.5*i)+1):
            if i % j ==0:
                is_prime=False
                break
        if is_prime:
            prime_list.append(i)

    return prime_list
starting_range = int(input())
ending_range = int(input())
lst = prime(starting_range, ending_range)
if len(lst) == 0:
    print("There are no prime numbers in this range")
else:
    print("The prime numbers in this range are:", lst)

2
7
The prime numbers in this range are: [2, 3, 5]


In [ ]:
def generate(num_rows):
    # Initialize the triangle list
    triangle = []
    for row_num in range(num_rows):
        row = [None for _ in range(row_num + 1)]
        row[0] = row[-1] = 1
        for j in range(1, len(row) - 1):
            row[j] = triangle[row_num - 1][j - 1] + triangle[row_num - 1][j]
        triangle.append(row)
    return triangle
rows = int(input())
result = generate(rows)
# Display the generated Pascal's Triangle
print(result)


5
[[1], [1, 1], [1, 2, 1], [1, 3, 3, 1], [1, 4, 6, 4, 1]]


###1. Excel Sheet Column Number
Medium
Given a string columnTitle that represents the column title as appears in an Excel sheet, return its corresponding column number.

John, a data analyst, was tasked with analyzing sales data and representing it in an Excel sheet. He decided to use alphabets as column titles. While assigning column numbers, he got stuck after column 26 and named it "Z". He then named the 27th column "AB" as the combination of alphabets.

Can you help John write a function to return the corresponding column number when a columnTitle is given?


A -> 1
B -> 2
C -> 3
...
Z -> 26
AA -> 27
AB -> 28
...

Example 1:
Input:
A

Output:
1

Example 2:
Input:
ZY

Output:
701

Constraints:
- 1 <= columnTitle.length <= 7
- columnTitle consists only of uppercase English letters.
- columnTitle is in the range ["A", "FXSHRXW"]
All quest

In [ ]:
def titleToNumber(s):
    # Initialize the result
    result = 0

    # Calculate the length of the string
    n = len(s)

    # Iterate over each character and update the result
    for i in range(n):
        # Get the character at index i
        char = s[i]
        # Calculate the numeric value of the character
        num_value = ord(char) - ord('A') + 1
        # Update the result
        result = result * 26 + num_value

    # Return the final result
    return result

# Prompt for user input
st = input("Enter the column title: ")

# Call the function and display the result
print(titleToNumber(st))


In [ ]:
def titleToNumber(s):
    # Initialization
    result = 0
    alpha_map = {chr(i): i - 64 for i in range(65, 91)}  # Mapping A=1, B=2, ..., Z=26

    # Calculate the length of the string
    n = len(s)

    # Iterate over each character in reverse
    for i in range(n):
        # Calculate the position of the character from the end
        pos = n - 1 - i
        # Retrieve the current character
        cur_char = s[pos]
        # Increment result
        result += alpha_map[cur_char] * (26 ** i)

    # Return the final result
    return result

# Prompt for user input
st = input("Enter the column title: ")

# Call the function and display the result
print(titleToNumber(st))


In [ ]:
def prime(x, y):
    prime_list = []
    for i in range(x, y):
        if i == 0 or i == 1:
            continue
        else:
            # Check if the number is prime
            # write your code here
            is_prime=True
            for j in range(2,int(i/2)+1):
                if i % j ==0:
                    is_prime=False
                break
            if is_prime:
                prime_list.append(i)

    return prime_list

# Driver program
x = int(input())
y = int(input())
lst = prime(x, y)
if len(lst) == 0:
    print("There are no prime numbers in this range")
else:
    print("The prime numbers in this range are:", lst)


2
11
The prime numbers in this range are: [2, 3, 5, 7, 9]


In [ ]:
def prime(x, y):
    prime_list = []
    for i in range(x, y):
        if i == 0 or i == 1:
            continue
        else:
            # Check if the number is prime
            # write your code here
            is_prime=True
            for j in range(2,int(i/2)+1):
              if i%j==0:
                is_prime=False
              break
              prime_list.append(i)

    return prime_list

# Driver program
starting_range = int(input("Enter the starting range: "))
ending_range = int(input("Enter the ending range: "))
lst = prime(starting_range, ending_range)
if len(lst) == 0:
    print("There are no prime numbers in this range")
else:
    print("The prime numbers in this range are:", lst)


Enter the starting range: 2
Enter the ending range: 7
There are no prime numbers in this range


In [ ]:
def prime(x, y):
    prime_list = []
    for i in range(x, y):
        if i == 0 or i == 1:
            continue
        else:
            # Check if the number is prime
            is_prime = True
            for j in range(2, int(i / 2) + 1):
                if i % j == 0:
                    is_prime = False
                    break
            if is_prime:
                prime_list.append(i)

    return prime_list

# Driver program
starting_range = int(input("Enter the starting range: "))
ending_range = int(input("Enter the ending range: "))
lst = prime(starting_range, ending_range)
if len(lst) == 0:
    print("There are no prime numbers in this range")
else:
    print("The prime numbers in this range are:", lst)


Enter the starting range: 2
Enter the ending range: 7
The prime numbers in this range are: [2, 3, 5]


In [ ]:
def isHappy(n):
    # Function to calculate the sum of squares of digits
    def get_next(number):
        # Calculate the sum of the squares of the digits
        # write your code here
        total_sum=0
        while number>0:
            digit=number%10
            total_sum=digit**2
            number//=10
        return total_sum

    # Track the numbers we've seen
    seen = set()

    # Loop to determine if the number is happy
    while n != 1 and n not in seen:
        # Add the number to the set and calculate the next number
        # write your code here
        seen.add(n)
        n=get_next(n)


    # Return True if the number is happy, False otherwise
    return n == 1

# Prompt for user input
n = int(input())

# Display whether the number is happy or not
if isHappy(n):
    print("True")
else:
    print("False")



In [ ]:
def isHappy(n):
    # Helper function to calculate the sum of squares of digits
    def get_next(number):
        total_sum = 0
        while number > 0:
            digit = number % 10  # Get the last digit
            total_sum += digit * digit  # Add square of the digit to total_sum
            number //= 10  # Remove the last digit
        return total_sum

    # Track the numbers we've seen to detect cycles
    seen = set()

    # Loop to determine if the number is happy
    while n != 1 and n not in seen:
        seen.add(n)  # Add the current number to the set of seen numbers
        n = get_next(n)  # Calculate the next number

    # Return True if the number is happy (i.e., it reached 1), False otherwise
    return n == 1

# Prompt for user input
n = int(input())

# Display whether the number is happy or not
print(isHappy(n))

19
True


In [ ]:
def prime(start, end):
    prime_list = []
    for num in range(start, end):
        if num > 1:
            is_prime = True
            for i in range(2, int(num ** 0.5) + 1):
                if num % i == 0:
                    is_prime = False
                    break
            if is_prime:
                prime_list.append(num)
    return prime_list

# Driver program
starting_range = int(input("Enter the starting range: "))
ending_range = int(input("Enter the ending range: "))
lst = prime(starting_range, ending_range)
if len(lst) == 0:
    print("There are no prime numbers in this range")
else:
    print("The prime numbers in this range are:", lst)


Enter the starting range: 2
Enter the ending range: 7
The prime numbers in this range are: [2, 3, 5]


In [1]:
def isValid(s):
    # Define matching pairs for brackets
    matching_bracket = {')': '(', '}': '{', ']': '['}
    stack = []

    # Process each character to check validity
    for char in s:
        if char in matching_bracket.values():
            # If the character is an opening bracket, push it onto the stack
            stack.append(char)
        elif char in matching_bracket.keys():
            # If the character is a closing bracket
            if stack and stack[-1] == matching_bracket[char]:
                # Pop the top of the stack if it matches the expected opening bracket
                stack.pop()
            else:
                # If it doesn't match or stack is empty, return False
                return False
        else:
            # If the character is not a bracket, ignore it
            continue

    # Return True if stack is empty, indicating valid string
    return len(stack) == 0

# Prompt for user input
s = input("Enter a string of brackets: ")

# Display if the string is valid or not
print("The string is valid." if isValid(s) else "The string is invalid.")


Enter a string of brackets: ()
The string is valid.


In [2]:
word = "testing"
length = len(word)
pattern = [[' ']*length for _ in range(length)]

for i in range(length):
    pattern[i][i] = word[i]
    pattern[i][length - i - 1] = word[i]

for row in pattern:
    print(' '.join(row))

t           t
  e       e  
    s   s    
      t      
    i   i    
  n       n  
g           g


In [4]:
def print_x_pattern(word):
    for i in range(len(word)):
        for j in range(len(word)):
            if i == j or i + j == len(word) - 1:
                print(word[j], end="")
            else:
                print(" ", end="")
        print()

print_x_pattern("testing")

t     g
 e   n 
  s i  
   t   
  s i  
 e   n 
t     g


In [6]:
def print_cross_pattern(size):
    for i in range(size):
        for j in range(size):
            if i == j or i + j == size - 1:
                print('*', end='')
            else:
                print(' ', end='')
        print()  # Move to the next line after each row

# User Input and Output
size = int(input())

# Call the function to draw the cross pattern
print_cross_pattern(size)


5
*   *
 * * 
  *  
 * * 
*   *


In [7]:
def print_triangle_pattern(n):
    for i in range(n):
        for j in range(n - i, 0, -1):
            if i % 2 == 0:
                print('1', end='')
            else:
                print('0', end='')
        print()  # Move to the next line after each row

# User Input and Output
try:
    n = int(input("Enter the size of the triangle pattern: "))
    if n <= 0:
        raise ValueError("Size must be a positive integer.")
    print_triangle_pattern(n)
except ValueError as ve:
    print("Invalid input:", ve)


Enter the size of the triangle pattern: 5
11111
0000
111
00
1


In [5]:
s="Hello World"
s=s[6:]
print(s)

World
